hugr-passes/src/const_fold.rs

```diff
@@ -1,221 +1,294 @@
+            let cst = hugr.add_node_with_parent(parent, Const::new(v));
+            let lcst = hugr.add_node_with_parent(parent, LoadConstant { datatype });
+            hugr.connect(cst, OutgoingPort::from(0), lcst, IncomingPort::from(0));
+            hugr.disconnect(n, inport);
+            hugr.connect(lcst, OutgoingPort::from(0), n, inport);
+        }
+        for n in remove_nodes {
+            hugr.remove_node(n);
+        }
+        Ok(())
+    }
+
+    /// Run the pass using this configuration
     pub fn run<H: HugrMut>(
         &self,
         hugr: &mut H,
         reg: &ExtensionRegistry,
     ) -> Result<(), ConstFoldError> {
         self.validation
-            .run_validated_pass(hugr, reg, |hugr: &mut H, _| {
-                loop {
-                    // We can only safely apply a single replacement. Applying a
-                    // replacement removes nodes and edges which may be referenced by
-                    // further replacements returned by find_consts. Even worse, if we
-                    // attempted to apply those replacements, expecting them to fail if
-                    // the nodes and edges they reference had been deleted,  they may
-                    // succeed because new nodes and edges reused the ids.
-                    //
-                    // We could be a lot smarter here, keeping track of `LoadConstant`
-                    // nodes and only looking at their out neighbours.
-                    let Some((replace, removes)) = find_consts(hugr, hugr.nodes(), reg).next()
-                    else {
-                        break Ok(());
-                    };
-                    hugr.apply_rewrite(replace)?;
-                    for rem in removes {
-                        // We are optimistically applying these [RemoveLoadConstant] and
-                        // [RemoveConst] rewrites without checking whether the nodes
-                        // they attempt to remove have remaining uses. If they do, then
-                        // the rewrite fails and we move on.
-                        if let Ok(const_node) = hugr.apply_rewrite(rem) {
-                            // if the LoadConst was removed, try removing the Const too.
-                            let _ = hugr.apply_rewrite(RemoveConst(const_node));
+            .run_validated_pass(hugr, reg, |hugr: &mut H, _| self.run_no_validate(hugr))
+    }
+
+    fn find_needed_nodes<H: HugrView>(
+        &self,
+        results: &AnalysisResults<ValueHandle, H>,
+        needed: &mut HashSet<Node>,
+    ) {
+        let mut q = VecDeque::new();
+        let h = results.hugr();
+        q.push_back(h.root());
+        while let Some(n) = q.pop_front() {
+            if !needed.insert(n) {
+                continue;
+            };
+
+            if h.get_optype(n).is_cfg() {
+                for bb in h.children(n) {
+                    //if results.bb_reachable(bb).unwrap() { // no, we'd need to patch up predicates
```

Comment:
Remove this comment (this is run before we create results anyway?)

Reply:
but i don't really understand the argument here -- wouldn't the commented code be better?